text/2022-01-22-inject-network-chaos-based-on-service.md

```diff
@@ -0,0 +1,85 @@
+# Title
+
+## Summary
+
+<!-- One para explanation of the proposal. -->
+Network chaos injection based on Kubernetes service is urgently needed.
+We did some research about ChaosMesh for a few weeks, also comparing with some competitor tools. The first impression is that ChaosMesh product is good, but not perfect, but we think it can do better.
+
+## Motivation
+
+<!-- Why are we doing this? What use cases does it support? What is the expected outcome? -->
+The IaaS of our company is completedly based on domestic multi-clouds, just like Ali Cloud, Tencent Cloud, and may have more in future. The target is to make our SaaS & PaaS deployed on Ali Cloud and Tencent Cloud at the same time, and keep both active and active, if one cloud is outage, and the other can be completedly hot backup so that we can reduce the MTTR as most as possible. 
+
+The architecture is briefly described as followed, the two clouds can be connected by fiber lines, and the key idea is to make SaaS & IaaS completely isolated, only the PaaS can be mutual accessed to exchange/sync data.
+
+```              
+┌┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┐
+┆               DNS                 ┆
+└┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┘
+  ∧                               ∧
+  |                               |
+  ∨                               ∨  
+┌┈┈┈┈┐                          ┌┈┈┈┈┐
+┆SaaS┆                          ┆SaaS┆
+└┈┈┈┈┘                          └┈┈┈┈┘
+  ∧                               ∧
+  |                               |
+  ∨                               ∨
+┌┈┈┈┈┐                          ┌┈┈┈┈┐
+┆PaaS┆ <---- fiber lines ---->  ┆PaaS┆
+└┈┈┈┈┘                          └┈┈┈┈┘
+  ∧                               ∧
+  |                               |
+  ∨                               ∨
+┌┈┈┈┈┐                          ┌┈┈┈┈┐
+┆IaaS┆                          ┆IaaS┆
+└┈┈┈┈┘                          └┈┈┈┈┘
+(Ali Cloud. cidr:172.25.0.0/16) (Tencent Cloud. cidr:10.32.0.0/16)
+```
+
+To make above great idea happened, we need to do more network chaos experiments routinely to check if the unitization deployment is still working or broken by any unexpected RD user changes. In Tencent Cloud K8S enviroment (172.25.0.0/16), like based on a specific namespace and service, we expect to inject network 100% loss for 10.32.0.0/16 as external targets, well, it works if the two pods ping directly, but it does not work while accessing Ali Cloud appname using service name. In fact, the service name is resolved to an IPVS LB, which is a proxy for the target appname.
```

Comment:
> In Tencent Cloud K8S enviroment (172.25.0.0/16),

It seems the CIDR is mismatched. :P

Reply:
> 

We have more CIDRs in our K8s environment, hope this mismatched info can't mess up my requirements.